test/parallel_api/numeric/numeric.ops/reduce_by_segment.pass.cpp

```diff
@@ -330,31 +330,32 @@ main()
         test_algo_four_sequences<ValueType, test_reduce_by_segment<BinaryPredicate, BinaryOperation>>();
 #    endif // TEST_DPCPP_BACKEND_PRESENT
 #endif     // !_PSTL_ICC_TEST_SIMD_UDS_BROKEN && !_PSTL_ICPX_TEST_RED_BY_SEG_OPTIMIZER_CRASH
-    }
+}
 
-    {
-        using ValueType = ::std::complex<float>;
-        using BinaryPredicate = UserBinaryPredicate<ValueType>;
-        using BinaryOperation = MaxAbsFunctor<ValueType>;
+template <template <typename T> typename BinaryPredicate, 
+          template <typename T> typename BinaryOperation>
+void
+run_test()
+{
+    run_test<int,    BinaryPredicate<int>,    BinaryOperation<int>>();
+    run_test<float,  BinaryPredicate<float>,  BinaryOperation<float>>();
+    run_test<double, BinaryPredicate<double>, BinaryOperation<double>>();
+}
 
+int
+main()
+{
 #if TEST_DPCPP_BACKEND_PRESENT
-        // Run tests for USM shared memory
-        test4buffers<sycl::usm::alloc::shared, test_reduce_by_segment<ValueType, BinaryPredicate, BinaryOperation>>();
-        // Run tests for USM device memory
-        test4buffers<sycl::usm::alloc::device, test_reduce_by_segment<ValueType, BinaryPredicate, BinaryOperation>>();
-        // test with flag pred
-        test_flag_pred<sycl::usm::alloc::device, class KernelName7, std::uint64_t>();
-        test_flag_pred<sycl::usm::alloc::device, class KernelName8, dpl::complex<float>>();
+    // test with flag pred
+    test_flag_pred<sycl::usm::alloc::device, class KernelName1, std::uint64_t>();
+    test_flag_pred<sycl::usm::alloc::device, class KernelName2, dpl::complex<float>>();
 #endif // TEST_DPCPP_BACKEND_PRESENT
 
-#if !_PSTL_ICC_TEST_SIMD_UDS_BROKEN && !_PSTL_ICPX_TEST_RED_BY_SEG_OPTIMIZER_CRASH
-#    if TEST_DPCPP_BACKEND_PRESENT
-        test_algo_four_sequences<test_reduce_by_segment<ValueType, BinaryPredicate, BinaryOperation>>();
-#    else
-        test_algo_four_sequences<ValueType, test_reduce_by_segment<BinaryPredicate, BinaryOperation>>();
-#    endif // TEST_DPCPP_BACKEND_PRESENT
-#endif     // !_PSTL_ICC_TEST_SIMD_UDS_BROKEN && !_PSTL_ICPX_TEST_RED_BY_SEG_OPTIMIZER_CRASH
-    }
+    run_test<::std::uint64_t,       UserBinaryPredicate<::std::uint64_t>,       MaxFunctor<::std::uint64_t>>();
+    run_test<::std::complex<float>, UserBinaryPredicate<::std::complex<float>>, MaxFunctor<::std::complex<float>>>();
+
+    run_test<::std::equal_to, ::std::plus>();
+    run_test<::std::equal_to, ::std::multiplies>();
```

Comment:
Could you double-check that there is not overflow?

Reply:
Yes: error exist regardless of `ONEDPL_WORKAROUND_FOR_IGPU_64BIT_REDUCTION` state, in both variants.

Reply:
I am not totally sure, but my rude estimation shows that there is overflow with multiplication:
```python
from functools import reduce

# check if that sequence may overflow
# T keys[n] = { 1, 2, 3, 4, 1, 1, 2, 2, 3, 3, 4, 4, 1, 1, 1, ...};
# T vals[n] = { n random numbers between 0 and 4 };

n = 100000
val_number = 4

# number of occurrences of one unique value
single_val_all = n/val_number
# calculate a longest sequence length
single_val_max_seq_n = 1
while single_val_all > 0:
    single_val_all -= single_val_max_seq_n
    single_val_max_seq_n += 1

average_value = reduce(lambda x, y: x + y, range(1, val_number))/val_number
multiply_reduction = average_value ** single_val_max_seq_n
plus_reduction = average_value * single_val_max_seq_n

print(f"possible max single sequence lenght: {single_val_max_seq_n} when there are {val_number} unique keys and n is {n}")
print(f"possible max single sequence reduction result (multiply): {multiply_reduction}")
print(f"possible max single sequence reduction result (plus): {plus_reduction}")
```
output:
```
possible max single sequence lenght: 225 when there are 4 unique keys and n is 100000
possible max single sequence reduction result (multiply): 4.1738158843886507e+39
possible max single sequence reduction result (plus): 337.5
```

Reply:
I guess it would be safer to limit `n` for multiply predicate, unless my estimation above is entirely wrong.

Reply:
Sorry, I have noticed that the data is generated in the range of `[0, 4]` (instead of `[1,4]`). Perhaps, there is no overflow error with `multiplies` as there are high chances to get `0`.

However, I would still recommend to adapt the data generation for `multiplies` predicate:  start with `1` and limit `n`. It can be done in a separate PR in order not to delay the workaround, while removing (or commenting-out) a case with `multiplies` in this PR.

Reply:
I think that this value generation
```C++
host_vals[i] = rand() % 5;
```
not corelated with initial problem for which this PR was created. And for `std::multiplies` with `int` too.

Reply:
My point was that the data generation should be adapted for reduction with `multiplies` predicate, which is introduced by this PR.

The current generation presumably results in overflows.
It may result in issues or not. I do not know if it is UB either. However, it look like a case which we do not want to test. That's a caveat that the PR may add another issue into the testing.

Anyway, it is a minor point. It affects only tests and it is an estimation. I suggest temporary removing `multiplies` case, until it is investigated better, or at least to add a note that it may result in UB for further investigation/fixing.

Reply:
So the root issue was in CMakeLists.txt. @mmichel11 will fix it.
I propose to save data generation in this test as is.

Reply:
I guess we are talking about different problems. The PR adds a workaround and more checks. That's great, thanks. 
The tests now also pass (after a correction in CMakeLists.txt). This is good too.

But it also introduces a case which may potentially backfire somewhen.
This note may alleviate difficulties with searching in case of something goes wrong.

```suggestion
    // TODO: investigate possible overflow
    run_test<::std::equal_to, ::std::multiplies>();
```